src/middlewared/middlewared/plugins/bootenv.py

```diff
@@ -48,5 +43,55 @@ def set_attribute(self, oid, attrs):
         clone = Update.FindClone(oid)
         return Update.CloneSetAttr(clone, **attrs)
 
+    @accepts(Dict(
+        'bootenv_create',
+        Str('name', required=True),
+        Str('source'),
+    ))
+    def do_create(self, data):
+
+        verrors = ValidationErrors()
+        self._clean_be_name(verrors, 'bootenv_create', data['name'])
+        if verrors:
+            raise verrors
+
+        kwargs = {}
+        source = data.get('source')
+        if source:
+            kwargs['bename'] = source
+        clone = Update.CreateClone(data['name'], **kwargs)
+        if clone is False:
+            raise CallError('Failed to create boot environment')
+        return data['name']
+
+    @accepts(Str('id'), Dict(
+        'bootenv_update',
+        Str('name', required=True),
+    ))
+    def do_update(self, oid, data):
+
+        verrors = ValidationErrors()
+        self._clean_be_name(verrors, 'bootenv_update', data['name'])
+        if verrors:
+            raise verrors
+
+        if not Update.RenameClone(oid, data['name']):
+            raise CallError('Failed to update boot environment')
+        return data['name']
+
+    def _clean_be_name(self, verrors, schema, name):
```

Comment:
This can be implemented as validator on `name`

Reply:
I am skeptical about implementing logical validators in the schema. I think we should restrict ourselves to direct schema validation, not anything as complex as running system calls etc.

Or, do you mean just the allowed chars?

Reply:
@william-gr yes, I've meant first check. Subsequent check could also be implemented as second validator, but since it contains specific and non-reusable code (that should also be async, shouldn't it?), I agree with you that it should stay where it is.

Reply:
@themylogin please check now